Remotion/Data/DomainObjects/Persistence/Rdbms/DbCommandBuilders/QueryDbCommandBuilder.cs

```diff
@@ -28,51 +31,72 @@ namespace Remotion.Data.DomainObjects.Persistence.Rdbms.DbCommandBuilders
-          var queryParameterValueAsString = (string)queryParameterValue;
+      var regexKey = SqlDialect.GetParameterName("regex");
+      var paramRegex = s_regexDictionary.GetOrAdd(
+          regexKey,
+          key => new Regex(Regex.Escape(key).Replace("regex", "\\b[\\w\\d_]+\\b"), RegexOptions.IgnoreCase | RegexOptions.Multiline));
 
-          statement = statement.Replace(parameterWithDefinition.QueryParameter.Name, queryParameterValueAsString);
-        }
+      foreach (Match match in paramRegex.Matches(Statement))
+      {
+        statementBuilder.Append(Statement.Substring(statementIndex, match.Index - statementIndex));
+        statementIndex = match.Index + match.Length;
+        var param = ParametersWithDefinition.SingleOrDefault(p => p.QueryParameter.Name.Equals(match.Value, StringComparison.OrdinalIgnoreCase));
+        if (param.QueryParameter != null)
+          statementBuilder.Append(GetParameterTextRepresentation(param));
         else
-        {
-          var dataParameterName = parameterWithDefinition.QueryParameter.Name;
-          var dataParameterValue = parameterWithDefinition.DataParameterDefinition.GetParameterValue(queryParameterValue);
-          var parameter = parameterWithDefinition.DataParameterDefinition.CreateDataParameter(command, dataParameterName, dataParameterValue);
+          statementBuilder.Append(match.Value);
+      }
+      statementBuilder.Append(Statement.Substring(statementIndex, Statement.Length - statementIndex));
+
+      foreach (var parameterWithDefinition in ParametersWithDefinition.Where(p => p.QueryParameter.ParameterType == QueryParameterType.Value))
+      {
+        var dataParameterName = parameterWithDefinition.QueryParameter.Name;
+        var dataParameterValue = parameterWithDefinition.DataParameterDefinition.GetParameterValue(parameterWithDefinition.QueryParameter.Value);
+        var parameter = parameterWithDefinition.DataParameterDefinition.CreateDataParameter(command, dataParameterName, dataParameterValue);
 
-          command.Parameters.Add(parameter);
-        }
+        command.Parameters.Add(parameter);
       }
 
-      command.CommandText = statement;
+      command.CommandText = statementBuilder.ToString();
       return command;
     }
+
+    protected virtual string GetParameterTextRepresentation (QueryParameterWithDataParameterDefinition parameterWithDefinition)
+    {
+      var queryParameterValue = parameterWithDefinition.QueryParameter.Value;
+      if (parameterWithDefinition.QueryParameter.ParameterType == QueryParameterType.Text)
+      {
+        Assertion.DebugAssert(
+            queryParameterValue is string,
+            "parameterWithType.QueryParameter.Value is string when parameterWithType.QueryParameter.ParameterType == Text");
+        return (string)queryParameterValue;
+      }
+
+      return parameterWithDefinition.QueryParameter.Name;
```

Comment:
it feels like we should either switch this would look better using IF/ELSE because both branches are equally important.